File: java/src/main/java/org/elections/Elections.java
package org.elections;

import java.text.DecimalFormat;
import java.util.*;

public class Elections {
    List<String> candidates = new ArrayList<>();
    List<String> officialCandidates = new ArrayList<>();
    ArrayList<Integer> votesWithoutDistricts = new ArrayList<>();
    Map<String, ArrayList<Integer>> votesWithDistricts;
    private Map<String, List<String>> list;
    private boolean withDistrict;

    public Elections(Map<String, List<String>> list, boolean withDistrict) {
        this.list = list;
        this.withDistrict = withDistrict;

        votesWithDistricts = new HashMap<>();
        votesWithDistricts.put("District 1", new ArrayList<>());
        votesWithDistricts.put("District 2", new ArrayList<>());
        votesWithDistricts.put("District 3", new ArrayList<>());
    }

    public void addCandidate(String candidate) {
        officialCandidates.add(candidate);
        candidates.add(candidate);
        votesWithoutDistricts.add(0);
        votesWithDistricts.get("District 1").add(0);
        votesWithDistricts.get("District 2").add(0);
        votesWithDistricts.get("District 3").add(0);
    }

    public void voteFor(String elector, String candidate, String electorDistrict) {
        if (!withDistrict) {
            if (candidates.contains(candidate)) {
                int index = candidates.indexOf(candidate);
                votesWithoutDistricts.set(index, votesWithoutDistricts.get(index) + 1);
            } else {
                candidates.add(candidate);
                votesWithoutDistricts.add(1);
            }
        } else {
            if (votesWithDistricts.containsKey(electorDistrict)) {
                ArrayList<Integer> districtVotes = votesWithDistricts.get(electorDistrict);
                if (candidates.contains(candidate)) {
                    int index = candidates.indexOf(candidate);
                    districtVotes.set(index, districtVotes.get(index) + 1);
                } else {
                    candidates.add(candidate);
                    votesWithDistricts.forEach((district, votes) -> {
                        votes.add(0);
                    });
                    districtVotes.set(candidates.size() - 1, districtVotes.get(candidates.size() - 1) + 1);
                }
            }
        }
    }

    public Map<String, String> results() {
        Map<String, String> results = new HashMap<>();
        Integer nbVotes = 0;
        Integer nullVotes = 0;
        Integer blankVotes = 0;
        int nbValidVotes = 0;

        if (!withDistrict) {
            nbVotes = votesWithoutDistricts.stream().reduce(0, Integer::sum);
            for (int i = 0; i < officialCandidates.size(); i++) {
                int index = candidates.indexOf(officialCandidates.get(i));
                nbValidVotes += votesWithoutDistricts.get(index);
            }

            for (int i = 0; i < votesWithoutDistricts.size(); i++) {
                Float candidatResult = ((float)votesWithoutDistricts.get(i) * 100) / nbValidVotes;
                String candidate = candidates.get(i);
                if (officialCandidates.contains(candidate)) {
                    results.put(candidate, String.format(Locale.FRENCH, "%.2f%%", candidatResult));
                } else {
                    if (candidates.get(i).isEmpty()) {
                        blankVotes += votesWithoutDistricts.get(i);
                    } else {
                        nullVotes += votesWithoutDistricts.get(i);
                    }
                }
            }
        } else {
            for (Map.Entry<String, ArrayList<Integer>> entry : votesWithDistricts.entrySet()) {
                ArrayList<Integer> districtVotes = entry.getValue();
                nbVotes += districtVotes.stream().reduce(0, Integer::sum);
            }

            for (int i = 0; i < officialCandidates.size(); i++) {
                int index = candidates.indexOf(officialCandidates.get(i));
                for (Map.Entry<String, ArrayList<Integer>> entry : votesWithDistricts.entrySet()) {
                    ArrayList<Integer> districtVotes = entry.getValue();
                    nbValidVotes += districtVotes.get(index);
                }
            }

            Map<String, Integer> officialCandidatesResult = new HashMap<>();
            for (int i = 0; i < officialCandidates.size(); i++) {
                officialCandidatesResult.put(candidates.get(i), 0);
            }
            for (Map.Entry<String, ArrayList<Integer>> entry : votesWithDistricts.entrySet()) {
                ArrayList<Float> districtResult = new ArrayList<>();
                ArrayList<Integer> districtVotes = entry.getValue();
                for (int i = 0; i < districtVotes.size(); i++) {
                    float candidateResult = 0;
                    if (nbValidVotes != 0)
                        candidateResult = ((float)districtVotes.get(i) * 100) / nbValidVotes;
                    String candidate = candidates.get(i);
                    if (officialCandidates.contains(candidate)) {
                        districtResult.add(candidateResult);
                    } else {
                        if (candidates.get(i).isEmpty()) {
                            blankVotes += districtVotes.get(i);
                        } else {
                            nullVotes += districtVotes.get(i);
                        }
                    }
                }
                int districtWinnerIndex = 0;
                for (int i = 1; i < districtResult.size(); i++) {
                    if (districtResult.get(districtWinnerIndex) < districtResult.get(i))
                        districtWinnerIndex = i;
                }
                officialCandidatesResult.put(candidates.get(districtWinnerIndex), officialCandidatesResult.get(candidates.get(districtWinnerIndex)) + 1);
            }
            for (int i = 0; i < officialCandidatesResult.size(); i++) {
                Float ratioCandidate = ((float) officialCandidatesResult.get(candidates.get(i))) / officialCandidatesResult.size() * 100;
                results.put(candidates.get(i), String.format(Locale.FRENCH, "%.2f%%", ratioCandidate));
            }
        }

        float blankResult = ((float)blankVotes * 100) / nbVotes;
        results.put("Blank", String.format(Locale.FRENCH, "%.2f%%", blankResult));

        float nullResult = ((float)nullVotes * 100) / nbVotes;
        results.put("Null", String.format(Locale.FRENCH, "%.2f%%", nullResult));

        int nbElectors = list.values().stream().map(List::size).reduce(0, Integer::sum);
        DecimalFormat df = new DecimalFormat();
        df.setMaximumFractionDigits(2);
        float abstentionResult = 100 - ((float) nbVotes * 100 / nbElectors);
        results.put("Abstention", String.format(Locale.FRENCH, "%.2f%%", abstentionResult));

        return results;
    }
}

File: kotlin/src/main/kotlin/org/elections/Elections.kt
package org.elections

import java.text.DecimalFormat
import java.util.*
import kotlin.collections.HashMap

class Elections(private val list: Map<String, List<String>>, private val withDistrict: Boolean) {

    var candidates: MutableList<String> = ArrayList()
    var officialCandidates: MutableList<String> = ArrayList()
    var votesWithoutDistricts = ArrayList<Int>()
    var votesWithDistricts: MutableMap<String, ArrayList<Int>>

    init {
        votesWithDistricts = HashMap()
        votesWithDistricts["District 1"] = ArrayList()
        votesWithDistricts["District 2"] = ArrayList()
        votesWithDistricts["District 3"] = ArrayList()
    }

    fun addCandidate(candidate: String) {
        officialCandidates.add(candidate)
        candidates.add(candidate)
        votesWithoutDistricts.add(0)
        votesWithDistricts["District 1"]!!.add(0)
        votesWithDistricts["District 2"]!!.add(0)
        votesWithDistricts["District 3"]!!.add(0)
    }

    fun voteFor(elector: String?, candidate: String, electorDistrict: String) {
        if (!withDistrict) {
            if (candidates.contains(candidate)) {
                val index = candidates.indexOf(candidate)
                votesWithoutDistricts[index] = votesWithoutDistricts[index] + 1
            } else {
                candidates.add(candidate)
                votesWithoutDistricts.add(1)
            }
        } else {
            if (votesWithDistricts.containsKey(electorDistrict)) {
                val districtVotes = votesWithDistricts[electorDistrict]!!
                if (candidates.contains(candidate)) {
                    val index = candidates.indexOf(candidate)
                    districtVotes[index] = districtVotes[index] + 1
                } else {
                    candidates.add(candidate)
                    votesWithDistricts.forEach { (district, votes) ->
                        votes.add(0)
                    }
                    districtVotes[candidates.size - 1] = districtVotes[candidates.size - 1] + 1
                }
            }
        }
    }

    fun results(): Map<String, String> {
        val results: MutableMap<String, String> = HashMap()
        var nbVotes = 0
        var nullVotes = 0
        var blankVotes = 0
        var nbValidVotes = 0
        if (!withDistrict) {
            nbVotes = votesWithoutDistricts.sum()
            for (i in officialCandidates.indices) {
                val index = candidates.indexOf(officialCandidates[i])
                nbValidVotes += votesWithoutDistricts[index]
            }
            for (i in votesWithoutDistricts.indices) {
                val candidatResult = votesWithoutDistricts[i].toFloat() * 100 / nbValidVotes
                val candidate = candidates[i]
                if (officialCandidates.contains(candidate)) {
                    results[candidate] = String.format(Locale.FRENCH, "%.2f%%", candidatResult)
                } else {
                    if (candidates[i].isEmpty()) {
                        blankVotes += votesWithoutDistricts[i]
                    } else {
                        nullVotes += votesWithoutDistricts[i]
                    }
                }
            }
        } else {
            for (districtVotes in votesWithDistricts.values) {
                nbVotes += districtVotes.sum()
            }
            for (i in officialCandidates.indices) {
                val index = candidates.indexOf(officialCandidates[i])
                for (districtVotes in votesWithDistricts.values) {
                    nbValidVotes += districtVotes[index]
                }
            }
            val officialCandidatesResult: MutableMap<String, Int> = HashMap()
            for (i in officialCandidates.indices) {
                officialCandidatesResult[candidates[i]] = 0
            }
            for (districtVotes in votesWithDistricts.values) {
                val districtResult = ArrayList<Float>()
                for (i in districtVotes.indices) {
                    var candidateResult = 0f
                    if (nbValidVotes != 0) candidateResult = districtVotes[i].toFloat() * 100 / nbValidVotes
                    val candidate = candidates[i]
                    if (officialCandidates.contains(candidate)) {
                        districtResult.add(candidateResult)
                    } else {
                        if (candidates[i].isEmpty()) {
                            blankVotes += districtVotes[i]
                        } else {
                            nullVotes += districtVotes[i]
                        }
                    }
                }
                var districtWinnerIndex = 0
                for (i in 1 until districtResult.size) {
                    if (districtResult[districtWinnerIndex] < districtResult[i]) districtWinnerIndex = i
                }
                officialCandidatesResult[candidates[districtWinnerIndex]] =
                    officialCandidatesResult[candidates[districtWinnerIndex]]!! + 1
            }
            for (i in 0 until officialCandidatesResult.size) {
                val ratioCandidate =
                    officialCandidatesResult[candidates[i]]!!.toFloat() / officialCandidatesResult.size * 100
                results[candidates[i]] = String.format(Locale.FRENCH, "%.2f%%", ratioCandidate)
            }
        }
        val blankResult = blankVotes.toFloat() * 100 / nbVotes
        results["Blank"] = String.format(Locale.FRENCH, "%.2f%%", blankResult)
        val nullResult = nullVotes.toFloat() * 100 / nbVotes
        results["Null"] = String.format(Locale.FRENCH, "%.2f%%", nullResult)
        val nbElectors = list.values.map { it.size }.sum()
        val df = DecimalFormat()
        df.maximumFractionDigits = 2
        val abstentionResult = 100 - nbVotes.toFloat() * 100 / nbElectors
        results["Abstention"] = String.format(Locale.FRENCH, "%.2f%%", abstentionResult)
        return results
    }
}
